app/Validation/ChangePasswordValidation.php

```diff
@@ -21,15 +26,55 @@ class ChangePasswordValidation extends BaseValidation
         ],
     ];
     protected $errors = [];
+    private $userWithToken;
+
+    public function __construct($userWithToken)
+    {
+        $this->userWithToken = $userWithToken;
+    }
 
-    public function existUserWithToken($user)
+    public function existUserWithToken()
     {
-        if (! $user) {
-            $this->errors = [
-                'connection' => 'Algo ha salido mal, porfavor recarga la página e intenta nuevamente',
-            ];
+        if (! $this->userWithToken) {
+            throw new Exception('Lo sentimos, el token proporcionado no es válido o no existe. Por favor, verifica la información e intenta nuevamente');
+        }
+
+        return true;
+    }
+
+    public function validateTokenExpiration($token)
+    {
+        $userTokenModelo = new UserTokenModel();
+
+        $tokenExpiryDate = $userTokenModelo->getTokenDate($token);
+        $today           = new DateTime();
+
+        if ($today >= $tokenExpiryDate) {
+            $responseDelete = $userTokenModelo->deleteToken($this->userWithToken['userId']);
+
+            if (! $responseDelete) {
+                throw new Exception('Lo sentimos, ha ocurrido un error inesperado. Por favor, inténtalo de nuevo.');
+            }
 
-            throw new Exception();
+            throw new Exception('El token ha expirado. Por favor, solicita uno nuevo para continuar con la operación.');
+        }
+
+        return true;
+    }
+
+    public function hasAccessToSetPassword()
+    {
+        if ($this->userWithToken['confirmed'] !== self::USER_IS_NOT_CONFIRMED) {
+            throw new Exception('Parece que no cuenta con el permiso para establecer su contraseña');
+        }
+
+        return true;
+    }
+
+    public function hasAccessToResetPassword()
+    {
+        if ($this->userWithToken['confirmed'] !== self::USER_IS_CONFIRMED) {
+            throw new Exception('Parece que no cuenta con el permiso para restablecer su contraseña');
```

Comment:
I have a doubt, in the `hasAccessToSetPassword()`, you need that user is confirmed right? then you can make a positive sentence on if statement, and in the `hasAccessToResetPassword()` function occurs similar, then:
```suggestion
    public function hasAccessToSetPassword()
    {
        if ($this->userWithToken['confirmed'] === self::USER_IS_CONFIRMED) {
            throw new Exception('Parece que no cuenta con el permiso para establecer su contraseña');
        }

        return true;
    }

    public function hasAccessToResetPassword()
    {
        if ($this->userWithToken['confirmed'] === self::USER_IS_NOT_CONFIRMED) {
            throw new Exception('Parece que no cuenta con el permiso para restablecer su contraseña');
```